docs/root/intro/arch_overview/matching/_include/complicated.yaml

```diff
@@ -0,0 +1,90 @@
+    filter_chains:
+      filters:
+      - name: envoy.filters.network.http_connection_manager
+        typed_config:
+          "@type": type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager
+          stat_prefix: ingress_http
+          http_filters:
+          - name: with-matcher
+            typed_config:
+              "@type": type.googleapis.com/envoy.extensions.common.matching.v3.ExtensionWithMatcher
+              extension_config:
+                  name: envoy.filters.http.fault
+                  typed_config:
+                    "@type": type.googleapis.com/envoy.extensions.filters.http.fault.v3.HTTPFault
+                    abort:
+                      http_status: 503
+                      percentage:
+                        numerator: 0
+                        denominator: HUNDRED
+                    delay:
+                      fixed_delay: 3s
+                      percentage:
+                        numerator: 0
+                        denominator: HUNDRED
+              matcher:
+                matcher_tree:
+                  input:
+                    name: request-headers
+                    typed_config:
+                      "@type": type.googleapis.com/envoy.type.matcher.v3.HttpRequestHeaderMatchInput
+                      header_name: some-header
+                  exact_match_map:
+                      # Note this additional indirection; this is a workaround for Protobuf oneof limitations.
+                      map:
+                        skip_filter: # This is the header value we're trying to match against.
+                          matcher:
+                            matcher_list:
+                              matchers:
+                              - predicate:
+                                  or_matcher:
+                                      predicate:
+                                      - single_predicate:
+                                          input:
+                                            name: request-headers
+                                            typed_config:
+                                              "@type": type.googleapis.com/envoy.type.matcher.v3.HttpRequestHeaderMatchInput
+                                              header_name: second-header
+                                          value_match:
+                                            exact: foo
+                                      - single_predicate:
+                                          input:
+                                            name: request-headers
+                                            typed_config:
+                                              "@type": type.googleapis.com/envoy.type.matcher.v3.HttpRequestHeaderMatchInput
+                                              header_name: second-header
+                                          value_match:
+                                            exact: bar
```

Comment:
A non-actionable comment that maybe we can just discuss thoughts on; the verbosity here is so high that I think we have two issues from a practical perspective:
1. Users who hand-write config are probably going to never want to use complex matchers. I feel the only way to use this reasonably is with some frontend language or API.
2. The config size is going to blow out at scale when we have type URL appearing everywhere. We've seen this also when looking at the impact of `xdstp://` URLs. It would be ironic if a solution aimed at scalability has a high cost due to config size explosion. I think for this problem we can maybe just add in some fixed matchers over time that take the place of common operations. This can be future work, but I think at scale it might be needed (e.g. a route table with 100k entries).

Reply:
For 1 I agree that people are probably never going to write complicated trees by hand: even for existing APIs most the advanced use cases are probably written at least using protobuf intellisense/autocomplete. Do you think this example ends up being too complicated to be useful?

I think this is a valid concern, though the config size explosion mostly matters at config update, not stream processing. At some point it can become an issue (CPU usage, delayed config application) but I don't think it would directly affect data plane performance. I wonder if it makes sense to reuse the pattern of static SDS secrets where we could define named extensions somewhere and then simply reference them by name in config instead of repeating the whole configuration. That would avoid the explosion in cases where the exact same matcher/input is used in a lot of different branches

Reply:
Not arguing against including the example, just sharing some food for thought. I agree the cost is only at config update, that's where we're seeing real CPU costs at O(10k) cluster or route configs. 

We've been thinking even about templating to make xdstp:// names work better, I'll keep this use case in mind as well when thinking through that.